core/framework/embedders.py

```
@@ -0,0 +1,997 @@
+                    # Last resort: raise a more informative error
+                    available_attrs = [attr for attr in dir(outputs) if not attr.startswith('_')]
+                    raise ValueError(f"Could not extract token embeddings from {type(outputs).__name__}. Available attributes: {available_attrs}")
+            
+            # Apply pooling or projection if needed to get to 2048 dims
+            # (Jina v4 uses a projection layer for final embeddings)
+            if hasattr(self.model, 'encode_text'):
+                # Use Jina's projection mechanism
+                # We need to pool to get document embedding first
+                pooled = token_embeddings.mean(dim=0, keepdim=True)
+                # Then project to final dimension
+                final_embedding = self.model.encode_text(
```

Ruff (0.12.2)

Local variable `final_embedding` is assigned to but never used

Remove assignment to unused variable `final_embedding`

(F841)